Add API reference template and guide (#252)
thegooddocsproject/templates: api-reference/api-reference-guide.md

# API reference guide

This `api-reference-guide` document provides extra writing tips describing **HOW** to fill in each of the sections within the [`api-reference-template`](api-reference-template.md) file.

- _Document version_: 0.1
- _Last updated_: June 27, 2022

## Overview

> Application Programme Interface (API) references are technical manuals that provide API specifications and integration instructions to help your audience understand the product. The audience can vary based on your product user groups, and they can be technical or non-technical.

Reviewer: nit: Application Programming Interface (API)

# API reference guide

This `api-reference-guide` document provides extra writing tips describing **HOW** to fill in each of the sections within the [`api-reference-template`](api-reference-template.md) file.

- _Document version_: 0.1
- _Last updated_: July 26, 2022

## Overview

Application Programme Interface (API) references are technical manuals that provide API specifications and integration instructions to help your audience understand the product. The audience can vary based on your product user groups, and they can be technical or non-technical.

> Accurate, concise, well-structured API documentation facilitates efficient adoption of APIs and increases the overall user experience.

Reviewer 2: **suggestion:** Would suggest: "improves the overall user experience" or "increases overall user engagement", depending on what was intended with the sentence.

The [`api-reference-template`](api-reference-template.md) is designed to help you build the API documentation efficiently and keep it consistent in both format and appearance.

## Document structure

> The template assumes that your API documentation set includes references for many API endpoints, that are organized into groups. The template includes three parts:

Reviewer 2: **question:** Is this assumption of "many API endpoints" needed? 

If someone has an API with just one or two endpoints, could they not benefit from using the template?

Reviewer 3: I second @csmeyns comment above --- I would cut this part entirely. Many APIs only have one endpoint. The template should be flexible enough to cover either one endpoint or many endpoints. Typically any grouping of endpoints is simply done by path and/or resource.

- [API overview section](#about-the-api-reference-overview-section) 
- [API resource reference section](#about-the-api-resource-reference-section) 
- [API endpoint reference section](#about-the-api-endpoint-reference-section)

> > add explanation about how to organize the sections. Change the H1 to H2. 

Reviewer 2: **question:** Was this a TODO that's not commented out, or meant to be part of the guide?

Reviewer 3: looks like it. this line should be cut.

## Before writing the API references

You may follow these guidelines to help you write better API reference documentation:

> - Familiarize yourself with the API protocol. If possible, try making some API calls in a testing environment.

Reviewer 2: **suggestion:** Here it could be helpful to give a link to or example of a possible API testing environment (e.g. Postman)?

Reviewer 3: There is no such thing as the API protocol. I would say something like "Familiarize yourself with how to to make requests to APIs." And yes +1 for recommending Postman, it's an easy way to get started making API requests without knowing how to code.

> - Adjust the organization of some sections to align with your API protocol. This API reference template works best with the [Representational State Transfer (REST)](https://en.wikipedia.org/wiki/Representational_state_transfer) API protocol.

Reviewer 2: **suggestion:** Perhaps the comment "This API reference (...) API protocol." could be placed earlier in the 'Overview' section instead? 

So that people who have something other than a REST API are forewarned

Reviewer 3: I would just say "This template is primarily for documenting REST APIs. Adjust as needed for other types of APIs."

- Interact with your API developers to learn about the APIs that you are documenting. Understand the data models and the logical connections between the API endpoints.
- Discuss the logic of grouping the API endpoints. Although grouping the endpoints by resource type is a common practice, as used in this template, it is also possible to group them by use case or other characteristics that better suit the user's needs.
- Conduct some user research about how the audience of your documentation would use the APIs. Identify the programming languages that your audience would most likely adopt to interact with your APIs.

> - Explore the possibility of auto-generating the API references.

Reviewer 2: **suggestion:** It could be helpful here to give an example of, of link to, tools that can be used for auto-generating API references?

> - Optimize the visual presentation of the API references by applying customized stylesheet, such as using syntax highlighting, table of contents, multi-column layout, etc.

Reviewer 4: **suggestion:** "applying a customized stylesheet"

## About the "API reference overview" section

> The API overview section in the API reference describes the API components which apply to all endpoints in the API. This section includes information on topics such as protocol, authorization method, versioning, status code.

Reviewer 2: ```suggestion
Use the API overview section in the API reference to describe the API components that apply to all endpoints in the API. In this section you can include information on topics such as protocol, authorization method, versioning, and status codes.
```

### About the "Introduction" section

> Use this section to provide a high-level overview of your API set, including the main features, communication protocol, content types, organization, etc.

Reviewer 2: ```suggestion
Use the Introduction section to provide a high-level overview of your API set, including the main features, communication protocol, content types, and organization.
```

### About the "Base URL" section

> Base URL is a common segment to which the API endpoint paths are appended. Defining the base URL in this section reduces the effort to duplicate the segment in each endpoint reference.

Reviewer 2: ```suggestion
The base URL is a common segment to which the API endpoint paths are appended. Defining the base URL in this section reduces the effort to duplicate the segment in each endpoint reference.
```

The base URL may differ for different groups of users or in different environments. If multiple base URLs are available, list them in this section and clearly describe the conditions of using each of them.

### About the "Authorization" section

Use this section to specify the authentication and authorization requirements of using the APIs, including the authorization type, request schema, possible error codes, and examples.

:information_source: **Tips**:

> - This section should only describe the technical specifications of the API authorization, such as parameters and token expiration. Details of how to authenticate should be documented in a separate how-to guide and be referred to as a link here.

Reviewer 2: ```suggestion
- Only use this section to describe the technical specifications of the API authorization, such as parameters and token expiration. Details of how to authenticate can be documented in a separate how-to guide and be referred to as a link here.
```

- Provide an example of the authorization request or command. You can use random strings to replace the actual secrets.
- If there is more than one authorization type applicable to your APIs, describe and specify the use case of each. If one of the options is preferred, highlight it.
- If different permission levels apply to different API endpoints, document the requirements in the reference for that endpoint.

### About the "Versioning" section

This section is optional. If applicable, use this section to describe the versioning, compatibility, and lifecycle policies of your API set. If migration guides are available, provide the links here.

### About the "Pagination" section

This section is optional. If applicable, use this section to describe the options and default values for dividing long API responses into pages.

> :information_source: **Tip**: Your readers might not understand what "pagination" is, so it is wise to embed an explanation in the descriptive text.

Reviewer 4: **suggestion:** "so embed an explanation..."

### About the "Rate limiting and throttling" section

This section is optional. If applicable, use this section to describe the global rate limiting settings and quota of your APIs. If different quotas apply to different endpoints, also add this section in the reference for endpoints.

### About the "HTTP status code" section

Use this section to provide the list of HTTP status code that are applicable to your APIs. It is possible that the APIs only return certain types of the standard HTTP status codes.

:information_source: **Tips**:

- For REST APIs, HTTP status codes are used. If you are writing references for other protocols, adjust the table columns.
- In the `Description` column, describe the causes of that error, and provide related information or links about how to deal with the error.

### About the "Errors" section

This section is optional. If applicable, use this section to list the custom error types defined for the APIs. For easier navigation, you may provide each of the error definitions as a subsection.

> ### Additional sections

Reviewer 2: **suggestion:** Perhaps other standard sections that could be included before the 'Additional sections' would be:
- License (specifying the license under which the API is released)
- Contact (specifying a contact for questions about the API)

If other specifications are applicable to the whole API set, such as data conventions, request retry logic, logging, monitoring, define your own sections here.

## About the "API resource reference" section

> The API resource reference section provides reference information for a subset of API endpoints that are grouped around a resource type. This section describes the data model of the resource type.

Reviewer 2: ```suggestion
Use the API resource reference section to provide reference information for a subset of API endpoints that are grouped around a resource type. You can use this section to describe the data model of the resource type.
```

The template assumes that the API endpoints are grouped by the resource type that they are interacting with.

**Title**

> Substitute the document title `{Resource Name}` with the actual name. The resource name usually uses the same naming convention as in the source code.

Reviewer 4: **nitpick:** you use "Resource name" capitalisation below.

- Typically the name of an API endpoint consists of the operation type and the resource type. For example, an API endpoint that creates a User resource can be named `Create user`.
- As in the API resource reference, the naming convention should be consistent throughout your API documentation.
- Use the singular form of the resource name unless the endpoint is designed exclusively for a bulk operation, such as `List users`.

**Short description**

Provide a one-line description of what the API does. Starts with a verb. For example:

- For "get" operations: `Retrieves a {resource}.`
- For "list" operations: `Lists {resources}.`
- For "create" operations: `Creates a {resource}.` or `Inserts a {resource}.`
- For "update" operations: `Updates a {resource}.`
- For "delete" operations: `Removes a {resource}.` 

Ensure that the description here is consistent with that listed in the resource reference.

### About the "Endpoint" section

Use this section to define the API endpoint.

:information_source: **Tips**:

- Use fenced code blocks to make your code distinctive from other text blocks.
- Replace {METHOD} with the actual request method and capitalize all letters. For example, `POST`.

> - In the {request_url} segment, start with a slash character `/` and only provide the URL path (the segment after the hostname). The base URL can be omitted if you already document it in the API overview. For example, `/v2/users`.

Reviewer 4: **suggestion:** "The base URL can be omitted if you have already documented it..."
